design/thin-edge-core.md

```diff
@@ -0,0 +1,203 @@
+# Thin-edge Core Vision
+
+A typical IoT application has to address:
+
+* __sophisticated business-cases__, with applications involving several vendors, actors and users;
+* __very diverse integration needs__ over operating systems, protocols, data sources, processing tools, actuators, cloud end-points;
+* __constraint environments__ with low-resources devices, high exposure to security risks, prohibitive manual operations;
+* __particular non-functional requirements__ safety of devices operating machines in hazardous contexts,
+  scalability over large fleet of devices, long-term supports of scattered devices.
+  
+The forces behind these requirements are pushing in opposite directions.
+
+* On one side, to address the diversity of use-cases and integration requirements, one needs a large open set of features that can be easily extended and combined.
+  [NodeRed](https://nodered.org/) is the archetype of such tools facilitating connections of components.
+* On the other side, to meet resource and security constraints, one needs a minimal system specifically built for a task with cherry-picked components.
+  On the extreme side, this leads to the idea of [unikernel](https://en.wikipedia.org/wiki/Unikernel) with sealed, single-purpose software images.
+  
+What makes thin-edge unique is its approach to reconcile these two poles with two levels of building blocks.
+* To maximize flexibility and interoperability, thin-edge provides the tools to build an agent from independent executables that interact over MQTT using JSON messages.
```

Comment:
The project of thin-edge has been built around the idea of loose-coupling over MQTT. After the discussion you opened, I'm now convinced that this is not enough and that one also needs strong coupling leveraging Rust features. However, loose-coupling cannot be removed from thin-edge : this would be another project. And not only because some wants to use microservices but also and mainly to interact with tools provided by others and to connect with peripherical devices.

Reply:
The project of thin-edge has been built around the idea of loose-coupling over MQTT. After the discussion you opened, I'm now convinced that this is not enough and that one also needs strong coupling leveraging Rust features. However, loose-coupling cannot be removed from thin-edge : this would be another project. And not only because some wants to use microservices but also and mainly to interact with tools provided by others and to connect with peripherical devices.

Reply:
> However, loose-coupling cannot be removed from thin-edge : this would be another project. And not only because some wants to use microservices but also and mainly to interact with tools provided by others and to connect with peripherical devices.

Please point out how loose-coupling of external tools or peripheril devices is impacted by what I said.